### PR TITLE
Update Error message for `VersionNotFoundError` to handle Permission related issues better

### LINKED ISSUE #1768: Catch error for issues with reading/writing data to object stores (e.g. S3)
The original Discord discussion https://discord.com/channels/778216384475693066/989568455512555540

# Context
> williamc — 06/23/2022
Let's say I just cloned my kedro project repo to another machine, and its datasets are versioned and configured to use S3 for storage. If I try to run a pipeline that depends on those datasets I get the infamous kedro.io.core.VersionNotFoundError . Bucket has versions all the way up to 2022-06-07T22.04.39.460Z/ and the error says 2022-06-23T16.20.52.945Z. Is this the intended behavior? Thanks

# Solution
User report that this is resolved after he changes the permission to `s3:*` from `s3:*Object`. I suspect we need  `ListObjects` or `ListObjectsV2` only.

# What's the problem?
I expect it to throw an error like `Insufficient permissions to list objects`, these logic is likely handled by `fsspec` and the `glob_function` we pass into `VersionedDataSet`. Some more investigation is needed

# Action:
1. Show a better error like `Insufficient permissions to list objects` instead of `VersionedNotFound`? This is confusing because the user can see the files but `kedro` would complain the file isn't there. It's not clear currently if the error doesn't come from Kedro directly but from e.g. fsspec
2. Check what kind of Policy are needed for `kedro` to work? (p.s. `*` is definitely not a good suggestion!) and document this clearly

### PR DESCRIPTION
**Pull request by ankatiyar**
Signed-off-by: Ankita Katiyar [110245118+ankatiyar@users.noreply.github.com](mailto:110245118+ankatiyar@users.noreply.github.com)

## Description
Resolves #1768

## Development notes
Fresh PR to resolve the issue.
Closing https://github.com/kedro-org/kedro/pull/1809 in favour of this solution - `PermissionError` was being handled outside of Kedro in some cases. This PR is a workaround as the error message is insufficient and it's hard to pinpoint the source of error precisely in case of cloud storage.

### Proposed Solution
As a result, this PR modifies the error message for `VersionNotFoundError` to indicate that this error could be happening due to insufficient permission in case of cloud storage i.e. S3, GCS etc.

## Checklist

- [ ] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [ ] Added tests to cover my changes

<a href="https://gitpod.io/#https://github.com/kedro-org/kedro/pull/1881"></a>

### CODE REVIEW

**Comment by noklam** [comment]
@ankatiyar Don't forget to document how this decision was made!

File: tests/io/test_data_catalog.py [hunk 1]
```diff
@@ -652,3 +652,14 @@ def test_replacing_nonword_characters(self):
         assert "ds2_spark" in catalog.datasets.__dict__
         assert "ds3__csv" in catalog.datasets.__dict__
         assert "jalapeño" in catalog.datasets.__dict__
+
+    def test_no_version_cloud(self):
```
Comment (noklam) [inline tests/io/test_data_catalog.py:656]:
```suggestion
    def test_no_versions_with_cloud_protocol(self):
```

File: tests/io/test_data_catalog.py [hunk 2]
```diff
@@ -652,3 +652,14 @@ def test_replacing_nonword_characters(self):
...
+        """Check the error if no versions are available for load from cloud storage"""
+        version = Version(load=None, save=None)
+        ds = CSVDataSet("s3://bucket/file.csv", version=version)
```
Comment (noklam) [inline tests/io/test_data_catalog.py:659]:
```suggestion
        versioned_dataset = CSVDataSet("s3://bucket/file.csv", version=version)
```

File: tests/io/test_data_catalog.py [hunk 3]
```diff
@@ -652,3 +652,14 @@ def test_replacing_nonword_characters(self):
...
+        pattern = re.escape(
+            f"Did not find any versions for {ds} "
+            f"This could be due to insufficient permission."
+        )
+        with pytest.raises(DataSetError, match=pattern):
+            ds.load()
```
Comment (noklam) [inline tests/io/test_data_catalog.py:665]:
```python
    def test_no_versions(self, versioned_plot_writer):
        """Check the error if no versions are available for load."""
        pattern = r"Did not find any versions for MatplotlibWriter\(.+\)"
        with pytest.raises(DataSetError, match=pattern):
            versioned_plot_writer.load()
```

I found this test in one of our datasets test, can we follow the same pattern here? I think "rf" should be the same as escape.

**Review by noklam — COMMENTED** [review]
Thank you for sticking with it, I think this is a tough one as the first few issues. It's in a good shape already, most of my comments are around the styling.

A good tip is always check the existing codebase and see if there are something similar exist. I

**Review by MerelTheisenQB — COMMENTED** [review]
Just to double check that I understand this correctly: the error returned by `fsspec` isn't a `PermissionError` and that's why we cannot know for sure that there is something wrong with permission?

File: kedro/io/core.py [hunk 4]
```diff
@@ -538,9 +538,16 @@ def _fetch_latest_load_version(self) -> str:
         most_recent = next(
             (path for path in version_paths if self._exists_function(path)), None
         )
-
+        protocol = getattr(self, "_protocol", None)
         if not most_recent:
-            raise VersionNotFoundError(f"Did not find any versions for {self}")
+            if protocol in CLOUD_PROTOCOLS:
+                message = (
+                    f"Did not find any versions for {self} This could be "
```
[the remark this answers is not in the document]
Comment (noklam) [reply kedro/io/core.py:545]: @MerelTheisenQB Correct, it's possible to get an error with some other functions , but all datasets rely on `self._fs.glob` and it won't give you any error with permission issues.

It seems that this is the expected behavior of `glob`. For example, if you actually `ls` the system, you will get a permission error. However, this is a very expensive operation especially with object stores, thus I didn't like it.

File: tests/io/test_data_catalog.py (hunk 1 above)
Comment (ankatiyar) [reply tests/io/test_data_catalog.py]: Made the change in 211c449b68a52d3204dcb884b9af671d24753b1f

File: tests/io/test_data_catalog.py (hunk 2 above)
Comment (ankatiyar) [reply tests/io/test_data_catalog.py]: Made the change in 211c449b68a52d3204dcb884b9af671d24753b1f

File: tests/io/test_data_catalog.py (hunk 3 above)
Comment (ankatiyar) [reply tests/io/test_data_catalog.py]: I tried using "rf" but it doesn't seem to work.

File: kedro/io/core.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ankatiyar) [reply kedro/io/core.py]: Made the change in 211c449b68a52d3204dcb884b9af671d24753b1f

[the remark this answers is not in the document]
Comment (MerelTheisenQB) [reply kedro/io/core.py]: Thanks for the explanation @noklam 🙂

**Review by MerelTheisenQB — APPROVED** [review]
This was definitely a lot trickier than we thought, but this solution makes sense to me 👍 

Don't forget to add the change in the release notes.

**Review by AhdraMeraliQB — APPROVED** [review]
Nice work on this!! Thanks for sticking with it 😄

**Pull request merged**